### DISCUSSION TITLE
UMB in HMA

### DISCUSSION THREAD

**Discussion by stsp** [dosemu2/dosemu2#1774] [General]
I rewrote the heap management in
fdpp and disabled HMA by default.
Its no longer needed.
I also added a new option `$_umb_hma`
which should allow yet another 64K
of UMB space. But its currently only
supported by fdpp and is untested.
Andrew, could you please help with
testing that? The problem is that
`mem` can't be used, it needs to be
updated to support this. So the test
is needed to check if such UMB works.
In particular that it returns segment
0xffff when you allocate from it.
After we test that with fdpp, we may
try to support other DOSes as well.
xms.c:292 can be flipped to get a
potential support for non-fdpp but
this also needs to be tested.

**Comment by stsp** [comment]
OK, I think this is an unrealistic task.
So I removed that option.
So we only need tests for these HMA
funcs:
http://www.ctyme.com/intr/rb-4815.htm
http://www.ctyme.com/intr/rb-4816.htm

**Comment by stsp** [comment]
@bolle732 do you know if geos
can use HMA? If so - it should be
quite happy now.

**Comment by bolle732** [comment]
Hi Stas

Thanks you for asking. I'm not really sure, but I like to test it.

With the current devel at 48d55c571510bb2f6c50dcd9acf0543fb23047fa, I can't start DOSEmu2 anymore. I get the following error:

```
ERROR: SMALLOC: Out Of Memory on alloc, requested=1023
Total size: 16384
Available space: 971 (15413 used)
Largest free area: 971
Memory pool dump:
	area: 1024 bytes, used
	area: 14384 bytes, used
	area: 5 bytes, used
	area: 971 bytes, free

ERROR: lowmem_heap: OOM, size=1023
```

Is this a known problem or are you still in the middle of the change? Otherwise, I could open a bug report.

**Reply by stsp** [reply]
No, it not known.
Is it with `$_dos_up` on or off?
at what point of boot does it OOM?
I guess I just need to add more heap,
but I wonder why I don't see that
overflow.

**Reply by stsp** [reply]
Could you check if it is related
to some customizations, eg in
your userhook.sys?

**Comment by bolle732** [comment]
It seems, that the `sudo make install` did not update all files. My `/etc/dosemu/dosemu.conf` didn't contain `$_dos_up`. So I deleted it, renamed the folder `.dosemu` and rerun make install. After that, I could start DOSEmu2 again.

So, I had `$_dos_up` set to on as it's seems be the default.

Then, I got the crash when I tried to start PC/GEOS. No `dosemurc` or `userhook.sys` was in place. I followed your advice to change `$_dos_up` to off and this made PC/GEOS starting again. Together with the old `userhook.sys`, it runs fine again.

My test program shows 956'288 bytes of heap in PC/GEOS. With the previous setup, it was at 980'624 bytes. But that included the 0x50 heap_seg hack, the 3 SFTs loaded high and some options which I need to re-investigate...

I'll play tomorrow with the memory options. Let's see, where we'll land ;-)

**Reply by stsp** [reply]
> It seems, that the sudo make install did not update all files. My /etc/dosemu/dosemu.conf didn't contain $_dos_up.

This is not a problem.

> renamed the folder .dosemu and rerun make install. After that, I could start DOSEmu2 again.

Then we don't know what options
caused OOM. But I'll add some heap
anyway to avoid that in a future.

> Then, I got the crash when I tried to start PC/GEOS.

Same crash with OOM?

**Reply by stsp** [reply]
> My test program shows 956'288 bytes of heap in PC/GEOS.

It may be that not all fdpp patches
landed yet if you use LP builds.
Otherwise please try `DOS=HIGH,UMB`
in userhook.sys to see if maybe
it can't otherwise reach HMA.

**Comment by bolle732** [comment]
Sidenote: under DOSBOX, I'll have 976'352 bytes of heap...

**Comment by stsp** [comment]
OOM should now be fixed.

**Comment by bolle732** [comment]
Thanks a lot! Yes, I can confirm, that the OOM is fixed. PC/GEOS now runs fine when `$_dos_up` is set to on.

I've applied again my configuration and tweaks. Regarding the heap size in PC/GEOS, I could notice an increment from 980'624 bytes to 981'120 bytes. This ~500 bytes corresponds mostly to the increased free DOS conventional memory (653'184 vs 653'696 bytes).

Is there an easy way to enable and disable HMA in DOSEmu2? With this, I could see if the PC/GEOS heap size changes and in turn we know it's using HMA or not.

**Reply by stsp** [reply]
What are those tweaks?
I remember reducing DOS_PSP,
anything else?

**Comment by bolle732** [comment]
And yes, the current DOSEmu2 beats all regarding PC/GEOS heap size ;-)

**Comment by bolle732** [comment]
Is the option `$_umb_hma` still there? Can I set it to (on) or is that by default? Didn't found any reference in `dosemu.conf`.

**Reply by stsp** [reply]
No but I replaced it now with
`$_hma` option so that you can
disable HMA and see.

**Comment by bolle732** [comment]
Is this `$_hma` option with `(on)` and `(off)`?

Yes, `heap_seg = 0x50` in `src/plugin/fdpp/hooks.c` of DOSEmu2 and `DynAlloc("sft", 1, sizeof(sftheader) + 3 * sizeof(sft))` in `kernel/config.c` of FDPP.

**Reply by stsp** [reply]
> Is this $_hma option with (on) and (off)?

Its default value is always in dosemu.conf.
If you don't see it then the patch not there
and you need a git pull.
Please update also fdpp as in a nearby
thread Andrew finds HMA bugs there.

**Comment by bolle732** [comment]
So, I've updated DOSEmu2 und FDPP again. Setting `$_hma` to `off` or `on`doesn't make any difference regarding allocated heap space of PC/GEOS.

Then I checked the sources of PC/GEOS:
1. I found, that the GEOS kernel has the ability to stay in the HMA. Unfortunately, it seems that this option is set to FALSE for the PC platform. Would be an interessting test...
2. Using the XMS swap driver will allocate UMBs and eventually the HMA. From the code, I can't say for sure that HMA is used or not. And at runtime, there is no known easy way to get this information.

At least from the test with `$_hma`, it seems PC/GEOS isn't using HMA at the moment under DOSEmu2.

**Comment by bolle732** [comment]
Ahm, this "kernel in HMA" seems not really implemented :( There's only a empty macro to handle access to virtual segments.

**Comment by bolle732** [comment]
In the XMS swap driver, there is a test routine for the HMA. Maybe I find a time slice to analyze that later.
